In [1]:
import numpy as np
import pandas as pd
import json
pd.set_option('display.max_rows', None)

In [2]:
file_path = './data/data.csv'

df = pd.read_csv(file_path)

df.head()

projectName  Sum  Best Practices  Possible Errors  Variables  \
0  leetcode-the-hard-way  254               4                0          7   
1    osm-community-index  262              16                0          0   
2             zigbee-OTA  283               1                0          1   
3                escrcpy  334               1                0          1   
4      HostlistsRegistry  340               0                0          0   

   Node.js and CommonJS  Stylistic Issues  ECMAScript6  no-multi-assign  \
0                     0               243            0                0   
1                     0               238            8                0   
2                     0               281            0                0   
3                     0               332            0                0   
4                     0               340            0                0   

   no-empty  ...  no-compare-neg-zero  no-caller  computed-property-spacing  \
0         0  ...                    0          0                          0   
1         0  ...                    0          0                          0   
2         0  ...                    0          0                          0   
3         0  ...                    0          0                          0   
4         0  ...                    0          0                          0   

   no-new-func  prefer-destructuring  no-constant-condition  \
0            0                     0                      0   
1            0                     0                      0   
2            0                     0                      0   
3            0                     0                      0   
4            0                     0                      0   

   operator-assignment  no-buffer-constructor  no-useless-return  no-undefined  
0                    0                      0                  2             0  
1                    0                      0                  0             0  
2                    0                      0                  0             0  
3                    0                      0                  0             0  
4                    0                      0                  0             0  

[5 rows x 253 columns]

In [3]:
rules_categories_path = 'rulesCategory.json'
with open(rules_categories_path, 'r') as f:
    rule_category = json.load(f)
category_rules = {}

for idx, r_c in enumerate(rule_category):
    category = r_c['type']
    rules = r_c['rules']
    category_rules[category] = rules
    print(f'{idx} - {category} - totle rules: {len(rules)}')
    print(f'rules: {rules}')

0 - Possible Errors - totle rules: 34
rules: ['for-direction', 'getter-return', 'no-await-in-loop', 'no-compare-neg-zero', 'no-cond-assign', 'no-console', 'no-constant-condition', 'no-control-regex', 'no-debugger', 'no-dupe-args', 'no-dupe-keys', 'no-duplicate-case', 'no-empty', 'no-empty-character-class', 'no-ex-assign', 'no-extra-boolean-cast', 'no-extra-parens', 'no-extra-semi', 'no-func-assign', 'no-inner-declarations', 'no-invalid-regexp', 'no-irregular-whitespace', 'no-obj-calls', 'no-prototype-builtins', 'no-regex-spaces', 'no-sparse-arrays', 'no-template-curly-in-string', 'no-unexpected-multiline', 'no-unreachable', 'no-unsafe-finally', 'no-unsafe-negation', 'use-isnan', 'valid-jsdoc', 'valid-typeof']
1 - Best Practices - totle rules: 70
rules: ['accessor-pairs', 'array-callback-return', 'block-scoped-var', 'class-methods-use-this', 'complexity', 'consistent-return', 'curly', 'default-case', 'dot-location', 'dot-notation', 'eqeqeq', 'guard-for-in', 'no-alert', 'no-caller', 'no-

In [4]:
categories = list(df.columns.values)[:7]
df[categories].head()

projectName  Sum  Best Practices  Possible Errors  Variables  \
0  leetcode-the-hard-way  254               4                0          7   
1    osm-community-index  262              16                0          0   
2             zigbee-OTA  283               1                0          1   
3                escrcpy  334               1                0          1   
4      HostlistsRegistry  340               0                0          0   

   Node.js and CommonJS  Stylistic Issues  
0                     0               243  
1                     0               238  
2                     0               281  
3                     0               332  
4                     0               340

# Stylistic Issues

In [5]:
stylistic_issues_rules = category_rules['Stylistic Issues']
df_StylisticIssues = df[stylistic_issues_rules]
df_StylisticIssues.head()

array-bracket-newline  array-bracket-spacing  array-element-newline  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   block-spacing  brace-style  camelcase  capitalized-comments  comma-dangle  \
0              4            4          0                     0            54   
1              0            0          0                     0             1   
2              0            0          0                     0             3   
3              0           21          0                     0           221   
4              0            0          0                     0            15   

   comma-spacing  comma-style  ...  space-before-blocks  \
0              0            0  ...                    0   
1              1            0  ...                    0   
2              0            0  ...                    0   
3              0            0  ...                    0   
4              0            0  ...                    0   

   space-before-function-paren  space-in-parens  space-infix-ops  \
0                           16                0                0   
1                           10                0                0   
2                            5                0                0   
3                           77                0                0   
4                            0                0                0   

   space-unary-ops  spaced-comment  switch-colon-spacing  \
0                0               0                     0   
1                0               0                     0   
2                0               0                     0   
3                0               0                     0   
4                0               0                     0   

   template-tag-spacing  unicode-bom  wrap-regex  
0                     0            0           0  
1                     0            0           0  
2                     0            0           0  
3                     0            0           0  
4                     0            0           0  

[5 rows x 87 columns]

In [41]:
print('Top 20 stylistic issues total')
df_StylisticIssues.sum().sort_values(ascending=False)[:20]

Top 20 stylistic issues total


indent                         11958325
semi                            8455594
quotes                          6864384
comma-spacing                   6025453
space-infix-ops                 5316848
no-tabs                         3437990
key-spacing                     1812991
block-spacing                   1700079
space-before-function-paren     1349539
semi-spacing                    1224946
object-curly-spacing            1192147
quote-props                     1140544
space-in-parens                 1027761
space-before-blocks              865285
comma-dangle                     810352
no-trailing-spaces               378031
camelcase                        375491
no-mixed-spaces-and-tabs         227208
jsx-quotes                       200425
one-var                          183844
dtype: int64

In [24]:
df_StylisticIssues.apply(lambda col: col[col != 0].count()).sort_values(ascending=False)[:5]

semi                           385
space-before-function-paren    385
indent                         365
quotes                         348
comma-dangle                   344
dtype: int64

#### 1 indent: JS does not care about indentation, but code become less readable
#### 2 semi: JS engine can automatically add it (ASI), but it can sts be tricky
`
    return
    {
        name: 'ESLint'
    };
    
    return; // below will be unreachable
    {
        name: 'ESLint';
    }
`
#### 3 quotes: can be single quote, double quote, and backticks
#### 4 comma-spacing: readability
#### 5 space-infix-ops: e.g.var sum = 1 + 2;
#### 6 no-tabs: tab characters
#### 7 key-spacing: spacing between keys and values    e.g. {a : b} => {a: b}
#### 8 block-spacing
#### 9 space-before-function-paren
#### 10 semi-spacing: e.g. { a: b }
#### 11 object-curly-spacing
### 12 quote-props
### ...
### 15 comma-dangle: 810352 => Trailing commas in object literals are valid according to ECMAScript5/3, BUT IE8 will throw error when it encounters trailing commas in JavaScript




# Variables

In [25]:
variables_rules = category_rules['Variables']
df_varialbes = df[variables_rules]
df_varialbes.head()

init-declarations  no-catch-shadow  no-delete-var  no-label-var  \
0                  0                0              0             0   
1                  0                0              0             0   
2                  0                0              0             0   
3                  0                0              0             0   
4                  0                0              0             0   

   no-restricted-globals  no-shadow  no-shadow-restricted-names  no-undef  \
0                      0          0                           0         6   
1                      0          0                           0         0   
2                      0          0                           0         0   
3                      0          0                           0         1   
4                      0          0                           0         0   

   no-undef-init  no-undefined  no-unused-vars  no-use-before-define  
0              0             0               1                     0  
1              0             0               0                     0  
2              0             0               1                     0  
3              0             0               0                     0  
4              0             0               0                     0

In [40]:
print('Top 5 variables total')
df_varialbes.sum().sort_values(ascending=False)[:5]

Top 5 variables total


no-undef                      1180244
no-unused-vars                  52793
no-use-before-define             8094
no-shadow-restricted-names       1791
no-undef-init                    1596
dtype: int64

In [30]:
df_varialbes.apply(lambda col: col[col != 0].count()).sort_values(ascending=False)[:5]

no-undef                      367
no-unused-vars                281
no-use-before-define          126
no-undef-init                 108
no-shadow-restricted-names     45
dtype: int64

#### 1 no-undef: Any reference to an undeclared variable causes a warning, unless the variable is explicitly mentioned in a /*global ...*/ comment.
#### 2 no-unused-vars: not used var or func (PROBLEM)
#### 3 no-use-before-define: use before define (before ES6 it can)
#### 4 no-shadow-restricted-names: Value Properties of the Global Object (NaN, Infinity, undefined) as well as strict mode restricted identifiers eval and arguments are considered to be restricted names in JavaScript. 
#### 5 no-undef-init: init var to undefined (JS automatically set, so no required to set manually)

In [31]:
no_unused_vars = df_varialbes['no-unused-vars']
has_no_unused_vars = np.count_nonzero(no_unused_vars)
print(f'There are {has_no_unused_vars} of 400 projects ({has_no_unused_vars / 400})')
max_count = np.max(no_unused_vars)
print(f'The one with most are {max_count} in project cstimer')

There are 281 of 400 projects (0.7025)
The one with most are 8251 in project cstimer


In [33]:
df.iloc[310]['no-unused-vars']

8251

In [34]:
no_shadow_restricted_names = df_varialbes['no-shadow-restricted-names']
has_no_shadow_restricted_names = np.count_nonzero(no_shadow_restricted_names)
print(f'There are {has_no_shadow_restricted_names} of 400 projects ({has_no_shadow_restricted_names / 400})')

There are 45 of 400 projects (0.1125)


In [35]:
df[df['no-shadow-restricted-names'] > 0][['projectName', 'no-shadow-restricted-names']].sort_values(by='no-shadow-restricted-names', ascending=False).head()

projectName  no-shadow-restricted-names
381   kendo-ui-core                        1022
395  sequenceserver                         302
378          AVideo                         128
350            docs                          83
366  EspruinoWebIDE                          34

code use like:
`(function( window, undefined ) {}()`

# ECMAScript6

In [36]:
ECMAScript6_rules = category_rules['ECMAScript6']
df_ECMAScript6 = df[ECMAScript6_rules]
df_ECMAScript6.head()

arrow-body-style  arrow-parens  arrow-spacing  constructor-super  \
0                 0             0              0                  0   
1                 0             0              0                  0   
2                 0             0              0                  0   
3                 0             0              0                  0   
4                 0             0              0                  0   

   generator-star-spacing  no-class-assign  no-confusing-arrow  \
0                       0                0                   0   
1                       0                0                   0   
2                       0                0                   0   
3                       0                0                   0   
4                       0                0                   0   

   no-const-assign  no-dupe-class-members  no-duplicate-imports  ...  \
0                0                      0                     0  ...   
1                0                      0                     0  ...   
2                0                      0                     0  ...   
3                0                      0                     0  ...   
4                0                      0                     0  ...   

   prefer-numeric-literals  prefer-rest-params  prefer-spread  \
0                        0                   0              0   
1                        0                   0              0   
2                        0                   0              0   
3                        0                   0              0   
4                        0                   0              0   

   prefer-template  require-yield  rest-spread-spacing  sort-imports  \
0                0              0                    0             0   
1                0              0                    0             0   
2                0              0                    0             0   
3                0              0                    0             0   
4                0              0                    0             0   

   symbol-description  template-curly-spacing  yield-star-spacing  
0                   0                       0                   0  
1                   0                       0                   0  
2                   0                       0                   0  
3                   0                       0                   0  
4                   0                       0                   0  

[5 rows x 31 columns]

In [39]:
print('Top 5 ECMAScript6 (total)')
df_ECMAScript6.sum().sort_values(ascending=False)[:5]

Top 5 ECMAScript6 (total)


no-var                    998295
arrow-spacing             118724
prefer-const              100530
object-shorthand           61899
generator-star-spacing      2771
dtype: int64

In [38]:
df_ECMAScript6.apply(lambda col: col[col != 0].count()).sort_values(ascending=False)[:5]

object-shorthand          282
prefer-const              260
no-var                    259
arrow-spacing             107
generator-star-spacing     73
dtype: int64

#### 1 object-shorthand: style for object
#### 2 prefer-const: use const if no reassign
#### 3 no-var: suggest use const or let (block scope)
#### 4 arrow-spaceing
#### 5 generator-star-spacing

# Best Practices

In [42]:
Best_Practices_rules = category_rules['Best Practices']
df_Best_Practices_rules= df[Best_Practices_rules]
df_Best_Practices_rules.head()

accessor-pairs  array-callback-return  block-scoped-var  \
0               0                      0                 0   
1               0                      0                 0   
2               0                      0                 0   
3               0                      0                 0   
4               0                      0                 0   

   class-methods-use-this  complexity  consistent-return  curly  default-case  \
0                       0           0                  0      1             0   
1                       0           0                  0      0             0   
2                       0           0                  0      0             0   
3                       0           0                  0      0             0   
4                       0           0                  0      0             0   

   dot-location  dot-notation  ...  no-useless-return  no-void  \
0             0             0  ...                  2        0   
1             0             0  ...                  0        0   
2             0             0  ...                  0        0   
3             0             0  ...                  0        0   
4             0             0  ...                  0        0   

   no-warning-comments  no-with  prefer-promise-reject-errors  radix  \
0                    0        0                             0      0   
1                    0        0                             0      0   
2                    0        0                             0      0   
3                    0        0                             1      0   
4                    0        0                             0      0   

   require-await  vars-on-top  wrap-iife  yoda  
0              0            0          0     0  
1              0            0          0     0  
2              0            0          0     0  
3              0            0          0     0  
4              0            0          0     0  

[5 rows x 70 columns]

In [43]:
print('Top 5 Best Practices (total)')
df_Best_Practices_rules.sum().sort_values(ascending=False)[:5]

Top 5 Best Practices (total)


eqeqeq                   249116
no-sequences             219923
no-unused-expressions    212069
yoda                     189973
no-labels                178238
dtype: int64

In [44]:
df_Best_Practices_rules.apply(lambda col: col[col != 0].count()).sort_values(ascending=False)[:5]

eqeqeq                   253
no-unused-expressions    225
dot-notation             225
curly                    223
no-useless-escape        203
dtype: int64

#### 1 eqeqeq: use type-safe equality operators === or !===
#### 2 no-unused-expressions: An unused expression which has no effect on the state of the program indicates a logic error
#### 3 dot-notation: use . to get properties better than square-bracket
#### 4 curly: never omit curly braces around blocks
#### 5 no-useless-escape: Escaping non-special characters in strings, template literals, and regular expressions

# Possible Errors

In [45]:
Possible_Errors_rules = category_rules['Possible Errors']
df_Possible_Errors_rules= df[Possible_Errors_rules]
df_Possible_Errors_rules.head()

for-direction  getter-return  no-await-in-loop  no-compare-neg-zero  \
0              0              0                 0                    0   
1              0              0                 0                    0   
2              0              0                 0                    0   
3              0              0                 0                    0   
4              0              0                 0                    0   

   no-cond-assign  no-console  no-constant-condition  no-control-regex  \
0               0           0                      0                 0   
1               0           0                      0                 0   
2               0           0                      0                 0   
3               0           0                      0                 0   
4               0           0                      0                 0   

   no-debugger  no-dupe-args  ...  no-regex-spaces  no-sparse-arrays  \
0            0             0  ...                0                 0   
1            0             0  ...                0                 0   
2            0             0  ...                0                 0   
3            0             0  ...                0                 0   
4            0             0  ...                0                 0   

   no-template-curly-in-string  no-unexpected-multiline  no-unreachable  \
0                            0                        0               0   
1                            0                        0               0   
2                            0                        0               0   
3                            0                        0               0   
4                            0                        0               0   

   no-unsafe-finally  no-unsafe-negation  use-isnan  valid-jsdoc  valid-typeof  
0                  0                   0          0            0             0  
1                  0                   0          0            0             0  
2                  0                   0          0            0             0  
3                  0                   0          0            0             0  
4                  0                   0          0            0             0  

[5 rows x 34 columns]

In [46]:
print('Top 5 Possible Errors (total)')
df_Possible_Errors_rules.sum().sort_values(ascending=False)[:5]

Top 5 Possible Errors (total)


no-extra-parens                49169
no-prototype-builtins          11518
no-template-curly-in-string     6286
no-cond-assign                  5901
no-func-assign                  2929
dtype: int64

In [47]:
df_Possible_Errors_rules.apply(lambda col: col[col != 0].count()).sort_values(ascending=False)[:5]

no-prototype-builtins    200
no-extra-parens          137
no-cond-assign           122
no-empty                 104
no-extra-boolean-cast     93
dtype: int64

#### 1 no-prototype-builtins: In ECMAScript 5.1, Object.create was added, which enables the creation of objects with a specified [[Prototype]]. Object.create(null) is a common pattern used to create objects that will be used as a Map. This can lead to errors when it is assumed that objects will have properties from Object.prototype.
#### 2 no-extra-parens: restricts the use of parentheses to only where they are necessary
#### 3 no-cond-assign: mistype a comparison operator e.g.== to =
#### 4 no-empty:Empty block statements
#### 5 no-extra-boolean-cast

In [51]:
df_Possible_Errors_rules[df_Possible_Errors_rules['no-cond-assign'] > 0]['no-cond-assign'].sort_values(ascending=False)[:5]

355    301
365    294
386    282
393    280
376    266
Name: no-cond-assign, dtype: int64

In [54]:
df.iloc[355]

projectName                         jeesite4
Sum                                   376467
Best Practices                          7111
Possible Errors                          459
Variables                               2319
Node.js and CommonJS                       0
Stylistic Issues                      355333
ECMAScript6                            11245
no-multi-assign                            0
no-empty                                  15
no-irregular-whitespace                    0
no-unreachable                             5
no-useless-concat                          0
no-restricted-properties                   0
space-before-function-paren             8376
consistent-this                            0
new-parens                                 0
one-var-declaration-per-line               0
id-match                                   0
no-debugger                                0
no-sync                                    0
no-lonely-if                               0
sort-impor